src/utils.js

```diff
@@ -33,36 +38,42 @@ export function clone(target) {
  * Merge any number of objects into a provided object.
  * @private
  */
-export function merge() {
-  let copy = null
-  let subject = null
+export function merge(): Object {
+  var copy = {}
+  var first = copy
+  var dirty = false
 
-  for (var i = 0, len = arguments.length; i < len; i++) {
-    copy = copy || arguments[i]
-    subject = subject || copy
+  for (var i = arguments.length - 1; i >= 0; i--) {
+    var subject = arguments[i]
+
+    if (isObject(subject) === false) {
+      continue
+    }
 
-    var next = arguments[i]
+    if (first === copy) {
+      first = subject
+    }
 
-    for (var key in next) {
-      if (copy[key] !== next[key]) {
-        if (copy === subject) {
-          copy = clone(subject)
-        }
+    for (var key in subject) {
+      if (!dirty) {
+        dirty = key in first === false
+      }
 
-        copy[key] = next[key]
+      if (key in copy === false) {
+        copy[key] = subject[key]
       }
     }
   }
 
-  return copy
+  return dirty ? copy : first
 }
```

Comment:
this is super cool, my only comment (as per usual), is that it's a bit tough to reason over.
- I think `first` is a misleading name since it's actually set to be the _last_ argument being passed in (i think). Hard to think of a better name though...
- `if (first === copy) { first = subject }` is the check that `first === copy` here the same as checking that `first` hasn't been set to something yet? If so, I think if it would be more clear if you initialized `var first = null` and checked it against null. Unless I'm missing something about the copy equality.
- The following example for the setting of `dirty` is much easier to reason about in my mind, no hard feelings but if you have similar thoughts, here's what my pass would be:
```js
if (!dirty && !(key in first)) {
  dirty = true
}
```